### PR TITLE
Course-Ruby: Project-2 

### PR DESCRIPTION
**Pull request by LamiaSristy**
## my_enumerables
In this project, I have implemented my own version of Ruby’s enumerable methods.

 Enumerable Methods List:
- my_each
- my_each_with_index
- my_select
- my_all?
- my_any?
- my_none?
- my_count
- my_map
- my_inject
- multiply_els

## Built With

- Ruby
- VS Code

### CODE REVIEW

File: lib/enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+      to_a.my_each { |item| return false unless [item.class, item.class.superclass].include?(param) }
+    elsif param.class == Regexp
+      to_a.my_each { |item| return false unless item.match(param) }
+    elsif !block_given?
+      to_a.my_each { |item| return false if item.nil? || !item }
+    else
+      to_a.my_each { |item| return false unless yield item }
+    end
+    true
+  end
```
Comment (bertil291utn) [inline lib/enumerable.rb]: ## False when doesn't match the pattern
- Should return false if any of the collection doesn't match the pattern
```ruby
p [3, 3, 2].my_all?(3) 
# => expected: false
# => got : true
```

File: lib/enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+      to_a.my_each { |item| return true if [item.class, item.class.superclass].include?(param) }
+    elsif param.class == Regexp
+      to_a.my_each { |item| return true if item.match(param) }
+    elsif !block_given?
+      to_a.my_each { |item| return true unless item.nil? || !item }
+    else
+      to_a.my_each { |item| return true if yield item }
+    end
+    false
+  end
```
Comment (bertil291utn) [inline lib/enumerable.rb]: ## False when doesn't match the pattern
- Should return false if any of the collection doesn't match the pattern
```ruby
p %w[dog parrot].my_any?("cat")
# => expected: false
# => got : true
```

File: lib/enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+  # 6.my_none?
+  def my_none?(param = nil)
+    if block_given?
+      !my_any?(&Proc.new)
+    else
+      !my_any?(param)
+    end
+  end
```
Comment (bertil291utn) [inline lib/enumerable.rb:78]: ## Should return True
- Returns true only if none of the collection matches the pattern
```ruby
p [3, 4, 2].my_none?(5)
#=> expected: true
#=> got : false

```

File: lib/enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  # 9.my_inject
+  def my_inject(initial = nil, sym = nil)
+    if !block_given? && !sym.nil?
+      to_a.my_each { |item| initial = initial.nil? ? item : initial.send(sym, item) }
+    else
+      to_a.my_each { |item| initial = initial.nil? ? item : yield(initial, item) }
+    end
+    initial
+  end
+end
```
Comment (bertil291utn) [inline lib/enumerable.rb]: ## Should accept also no block given
- When pass no block should work with a proc, Follow this [sample](https://gist.github.com/bertil291utn/5a7e6a73351c38ceb9781a9a37b58cf6)

```ruby
# p [3, 4, 2].my_inject(:+)
#=> expected: 9
#=> got : no block given error
```

**Review by bertil291utn — CHANGES_REQUESTED** [review]
# **Status:** Changes requested 🟠 
<p align="center"> image <br><strong>Not yet</strong> </p>

# Hey guys, you've been doing a great job, and also I hope you enjoyed coding.

### Your work is amazing you master it GitHub Flow you have a professional README file. 

However, in order to accomplish all the requirements, you required to do these changes, listed below:
- My_inject method
- My_any?
- My_all?
### Summary

-  Don't forget to work with inline comments below this comment box.

***

Guys work in the requested changes above and request another review.
Good luck. 

> “Never stop, never conform until what is good becomes even better, and what’s better becomes excellent.”

File: lib/enumerable.rb (hunk 1 above)
Comment (LamiaSristy) [reply lib/enumerable.rb]: It's now working, I have changed the method.

File: lib/enumerable.rb (hunk 3 above)
Comment (LamiaSristy) [reply lib/enumerable.rb:78]: It's now working, I have changed the method.

File: lib/enumerable.rb (hunk 4 above)
Comment (LamiaSristy) [reply lib/enumerable.rb]: It's now working, I have changed the method.

File: lib/enumerable.rb (hunk 2 above)
Comment (LamiaSristy) [reply lib/enumerable.rb]: It's working now, I have changed the method.

File: lib/enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+      num = nil
+      to_a.my_each { |item| num = num.nil? ? item : num.send(initial, item) }
+      num
+    elsif !sym.nil? && (sym.is_a?(Symbol) || sym.is_a?(String))
+      num = initial
+      to_a.my_each { |item| num = num.nil? ? item : num.send(sym, item) }
+      num
+    end
+  end
+end
```
Comment (christianotieno) [inline lib/enumerable.rb]: # my_inject
Kindly let us make sure our my_inject method obeys the following remaining working order 👇 
- [x] Our method should be raising a `"LocalJumpError"` when no block or argument is given.

**Review by christianotieno — CHANGES_REQUESTED** [review]
# **CHANGES REQUIRED** :warning:
Hey @LamiaSristy  :wave:
-
You have done a splendid job with the changes! :tada: Keep it up.
We, however, do have some few changes to carry out before we proceed (see review comment).
##
####  1. Kindly implement the `multiply_els` method and make sure it works as explained below 👇 
- [x] It should accept an array as an argument and multiplies all the elements of the array together using `my_inject` method.

#### 2. Please find the rest of the review in the down below. Thank you. 
##
- If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media or email. :eyes:
##
- When you have made the specified changes, please submit another code review. Thank you!
Happy coding :smile:
**[Christian Otieno](https://github.com/ChristianOtieno)**
[Twitter](https://twitter.com/iamchrisotieno) [linkedin](https://www.linkedin.com/in/christianotieno/) [GitLab](https://gitlab.com/christianotieno)
---
*P.S. all requested changes not marked by [OPTIONAL] or are not positive comments are **REQUIRED***

File: lib/enumerable.rb (hunk 5 above)
Comment (LamiaSristy) [reply lib/enumerable.rb]: I have made the changes, required.

**Review by BrittanyBlake — APPROVED** [review]
### Status: Approved ✅ 
****
### Hi Team! 👋 
### Great job implementing the requested changes!! 🎖️ It's a 👍 from me! Go ahead and merge this pull request and keep up the good work 😄
****

- Please note that you have used incorrect **Git Flow** in this project.
    - When using **Git Flow** (which was required for this project), You should be working from a **feature branch** and your pull request should be merging from the **feature branch** into the **development branch**. When your project is approved, you will merge from the **feature branch** into the **development branch**. 
- As this was not marked as invalid in your first review there is no need to make any changes, however please ensure that you are using correct Git Flow in your future projects where it is a requirement.

 Good Luck and Happy Coding! 🍀 👩🏾‍💻

**Pull request merged**